geps/gep-3440/index.md

```diff
@@ -0,0 +1,281 @@
+It relies on HTTP-level conditions for retries, so its gRPC support is less granular than the GEP requires.
+2. **Retry Limits**: HAProxy allows you to set a maximum number of retries for a request using the `retries` directive.
+It also supports setting a timeout for the entire retry process using the `timeout connect` and `timeout server` directives.
+
+### Traefik
+1. **Retry Conditions**: Traefik allows for retries based on HTTP-level conditions (e.g., connection errors and
+certain HTTP status codes like 500, 502, 503, and 504), but it does not natively interpret specific gRPC error codes
+like `UNAVAILABLE` or `DEADLINE_EXCEEDED`. This means that, while Traefik can retry requests on common HTTP errors
+that might represent temporary issues, it lacks the ability to directly handle and retry based on
+gRPC-specific error codes, limiting its alignment with the GEP’s requirement for granular gRPC error handling.
+2. **Retry Limits**: Traefik provides configurable retry attempts and can set a maximum number of retries. However,
+Traefik does not offer per-try timeout controls specific to each retry attempt. Instead, it typically relies on a
+global request timeout, limiting the flexibility needed for more precise gRPC retry management (like Envoy’s `per_try_timeout`).
+
+## API
+Having a dedicated API for gRPC retry conditions is necessary because gRPC uses
+unique error codes (e.g., `UNAVAILABLE`, `DEADLINE_EXCEEDED`) that represent transient issues specific to its protocol,
+which are not adequately covered by general HTTP status codes. gRPC also supports streaming and real-time communications,
+making retry strategies more complex than those used for standard HTTP requests. Existing proxies like Envoy handle
+gRPC retries with specialized logic, while other proxies rely on HTTP error codes, lacking the precision needed
+for gRPC.
+
+### Go
+
+```go
+type GRPCRouteRule struct {
+    // Retry defines the configuration for when to retry a gRPC request.
+    //
+    // Support: Extended
+    //
+    // +optional
+    // <gateway:experimental>
+    Retry *GRPCRouteRetry `json:"retry,omitempty"`
+
+    // ...
+}
+
+// GRPCRouteRetry defines retry configuration for a GRPCRoute.
+//
+// Implementations SHOULD retry on common transient gRPC errors
+// if a retry configuration is specified.
+//
+type GRPCRouteRetry struct {
+    // Reasons defines the gRPC error conditions for which a backend request
+    // should be retried.
+    //
+    // Supported gRPC error conditions:
+    // * "cancelled"
+    // * "deadline-exceeded"
+    // * "internal"
+    // * "resource-exhausted"
+    // * "unavailable"
+    //
+    // Implementations MUST support retrying requests for these conditions
+    // when specified.
+    //
+    // Support: Extended
+    //
+    // +optional
+    // <gateway:experimental>
+    Reasons []GRPCRouteRetryCondition `json:"reasons,omitempty"`
```

Comment:
Why name this `condition`? The name the GRPC core uses is `status-code` https://grpc.io/docs/guides/status-codes/

Reply:
Thanks for pointing this out. I am not going to lie, I got influenced by Envoy here as well 😅 
Gonna change it